modules/ding_wayf_dk/modules/ding_wayf_dk_login/ding_wayf_dk_login.module

```diff
@@ -0,0 +1,255 @@
+        }
+      }
+    }
+    else {
+      drupal_set_message(t('Something went wrong. Please try again or !url of WAYF.', array('!url' => l(t('logout'), "wayf/logout"))), 'warning');
+    }
+  }
+
+  ding_wayf_dk_login_redirect_user();
+}
+
+/**
+ * Implements hook_wayf_dk_login_pre_authenticate().
+ *
+ * Store destination for the current user, so redirect to e.g. page you where
+ * looking at and not always the front page.
+ */
+function ding_wayf_dk_login_wayf_dk_login_pre_authenticate() {
+  if (isset($_REQUEST['destination'])) {
+    $_SESSION['ding_wayf_dk_login_destination'] = $_REQUEST['destination'];
+
+    // The WAYF authentication process will redirect the users browser to WAYF
+    // servers, so session will not be started. Hence we need to start the
+    // session to ensure that the destination above don't get lost in redirects.
+    drupal_session_start();
+  }
+}
+
+/**
+ * Implements hook_wayf_dk_login_pre_logout().
+ *
+ * Ensures that session is cleaned up.
+ */
+function ding_wayf_dk_login_wayf_dk_login_pre_logout() {
+  if (isset($_SESSION['ding_wayf_dk_login_destination'])) {
+    unset($_SESSION['ding_wayf_dk_login_destination']);
+  }
+}
+
+/**
+ * Implements hook_wayf_dk_login_authenticate_result_alter().
+ *
+ * Restore destination after all the WAYF redirects.
+ */
+function ding_wayf_dk_login_wayf_dk_login_authenticate_result_alter($result) {
+  if (isset($_SESSION['ding_wayf_dk_login_destination'])) {
+    $_REQUEST['destination'] = $_SESSION['ding_wayf_dk_login_destination'];
+  }
+}
+
+/**
+ * Implements hook_preprocess_wayf_dk_login().
+ *
+ * Change the image to Nem-id.
+ */
+function ding_wayf_dk_login_preprocess_wayf_dk_login(&$variables) {
+  $image = array(
+    '#theme' => 'image',
+    '#path' => drupal_get_path('module', 'ding_wayf_dk_login') . '/images/nemid_logo.png',
+    '#alt' => t('Login with NemID'),
+    '#title' => t('Login with NemID'),
+  );
+
+  $variables['login_url']['#text'] = t('Login with') . ' ' . drupal_render($image);
+}
+
+/**
+ * Helper function to redirect the user.
+ *
+ * If destination query string exists the user is redirect to it else the
+ * fallback URL given is used.
+ *
+ * @param string $fallback_url
+ *   URL used if the destination query sting is not set.
+ */
+function ding_wayf_dk_login_redirect_user($fallback_url = NULL) {
+  if (!empty($_REQUEST['destination'])) {
+    // We do not user drupal_get_destination here as it will set the destination
+    // if one do not exists.
+    $fallback_url = $_REQUEST['destination'];
+  }
+  else {
+    global $user;
+    $fallback_url = token_replace($fallback_url, array('user' => $user));
+  }
+
+  drupal_goto($fallback_url);
+}
+
+/**
+ * Implements hook_form_wayf_dk_login__settings_form_alter().
+ *
+ * Adds debug option to map test WAYF CPR number to a provider valid one.
+ */
+function ding_wayf_dk_login_form_wayf_dk_login__settings_form_alter(&$form, &$form_state, $form_id) {
```

Comment:
To have all related configuration in the same form. The admin page for WAYF is "tab" based.